include/envoy/secret/secret_provider.h

```diff
@@ -23,6 +23,16 @@ template <class SecretType> class SecretProvider {
    */
   virtual const SecretType* secret() const PURE;
 
+  /**
+   * Add secret validation callback into secret provider.
+   * It is safe to call this method by main thread and callback is safe to be invoked
+   * on main thread.
+   * @param callback callback that is executed by secret provider.
+   * @return CallbackHandle the handle which can remove that validation callback.
+   */
+  virtual Common::CallbackHandle* addValidationCallback(
+      std::function<void(const envoy::api::v2::auth::CertificateValidationContext&)> callback) PURE;
```

Comment:
Why is it envoy::api::v2::auth::CertificateValidationContext, not SecretType? This SecretProvider only fetches secret of a specified SecretType, it will call the callback the verify if that received secret could be initialized successfully. @lizan

Reply:
+1 this should be SecretType